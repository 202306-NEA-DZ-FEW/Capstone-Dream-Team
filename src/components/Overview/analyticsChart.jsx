import { onAuthStateChanged } from "firebase/auth";
import {
    collection,
    getDocs,
    onSnapshot,
    query,
    where,
} from "firebase/firestore";
import dynamic from "next/dynamic";
import React, { useEffect, useState } from "react";

import { auth, db } from "@/util/firebase";

const ReactApexChart = dynamic(() => import("react-apexcharts"), {
    ssr: false,
});

function ChartGraph() {
    const [authUser, setAuthUser] = useState(null);
    const [chartData, setChartData] = useState(null);
    const yearlyAnalyticsChartOptions = {};
    //   legend: {
    //     show: false,
    //     position: "top",
    //     horizontalAlign: "left",
    //   },
    //   colors: ["#3C50E0", "#80CAEE"],
    //   chart: {
    //     fontFamily: "Satoshi, sans-serif",
    //     height: 335,
    //     type: "area",
    //     dropShadow: {
    //       enabled: true,
    //       color: "#623CEA14",
    //       top: 10,
    //       blur: 4,
    //       left: 0,
    //       opacity: 0.1,
    //     },
    //     toolbar: {
    //       show: true,
    //     },
    //   },
    //   responsive: [
    //     {
    //       breakpoint: 1024,
    //       options: {
    //         chart: {
    //           height: 300,
    //         },
    //       },
    //     },
    //     {
    //       breakpoint: 1366,
    //       options: {
    //         chart: {
    //           height: 350,
    //         },
    //       },
    //     },
    //   ],
    //   stroke: {
    //     width: [2, 2],
    //     curve: "straight",
    //   },
    //   grid: {
    //     xaxis: {
    //       lines: {
    //         show: true,
    //       },
    //     },
    //     yaxis: {
    //       lines: {
    //         show: true,
    //       },
    //     },
    //   },
    //   dataLabels: {
    //     enabled: false,
    //   },
    //   markers: {
    //     size: 4,
    //     colors: "#fff",
    //     strokeColors: ["#3056D3", "#80CAEE"],
    //     strokeWidth: 3,
    //     strokeOpacity: 0.9,
    //     strokeDashArray: 0,
    //     fillOpacity: 1,
    //     discrete: [],
    //     hover: {
    //       size: undefined,
    //       sizeOffset: 5,
    //     },
    //   },
    //   xaxis: {
    //     type: "category", // Change type to "category" for weekdays
    //     categories: ["Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"], // Weekdays as categories
    //     axisBorder: {
    //       show: false,
    //     },
    //     axisTicks: {
    //       show: false,
    //     },
    //   },
    //   yaxis: {
    //     title: {
    //       style: {
    //         fontSize: "0px",
    //       },
    //     },
    //     min: 0,
    //     max: 5,
    //   },
    // };

    useEffect(() => {
        onAuthStateChanged(auth, (user) => {
            user ? setAuthUser(user) : setAuthUser(null);
        });

        const fetchInformation = async () => {
            if (authUser) {
                const userId = authUser.uid;
                const q = query(
                    collection(db, "donors"),
                    where("restaurantId", "==", userId)
                );

                const querySnapshot = await getDocs(q);
                const chart = {};

                querySnapshot.forEach((doc) => {
                    const meal = doc.data();
                    const date = meal.date;

                    if (chart[date]) {
                        chart[date]++;
                    } else {
                        chart[date] = 1;
                    }
                });

                const chartArray = Object.keys(chart).map((date) => ({
                    date: date,
                    count: chart[date],
                }));

                // Add a null check before mapping
                if (chartArray && chartArray.length > 0) {
                    const chartData = {
                        labels: chartArray.map((data) => data.date),
                        datasets: [
                            {
                                label: "Donation per day",
                                data: chartArray.map((data) => data.count),
                                borderColor: "rgb(255, 99, 132)",
                                backgroundColor: "rgba(255, 99, 132, 0.5)",
                            },
                        ],
                    };
                    setChartData(chartData);
                } else {
                    // Handle the case when chartArray is null or empty
                    setChartData(null);
                }
            }
        };

        const unsubscribe = onSnapshot(collection(db, "donors"), () => {
            fetchInformation();
        });

        return () => {
            unsubscribe();
        };
    }, [authUser]);

    return (
        <div className='col-span-12 rounded-sm border border-stroke bg-white px-5 pt-7 pb-7 shadow sm:px-7.5 xl:col-span-8'>
            <div className='flex w-full flex-col flex-wrap items-start justify-between gap-3 sm:flex-nowrap'>
                <p className='font-bold text-primary'>Donation per day </p>
                <div className='w-full'>
                    <div id='YearlyAnalyticsChart' className='-ml-5'>
                        {chartData && (
                            <ReactApexChart
                                options={yearlyAnalyticsChartOptions}
                                series={[
                                    {
                                        name: "Donation per day",
                                        data: chartData
                                            ? chartData.datasets[0].data
                                            : [],
                                    },
                                ]}
                                type='area'
                                height={350}
                            />
                        )}
                    </div>
                </div>
            </div>
        </div>
    );
}

export default ChartGraph;
