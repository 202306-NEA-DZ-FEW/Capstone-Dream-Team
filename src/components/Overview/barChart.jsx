import {
    BarElement,
    CategoryScale,
    Chart,
    Legend,
    LinearScale,
    LineElement,
    PointElement,
    TimeScale,
    Title,
    Tooltip,
} from "chart.js";
import { onAuthStateChanged } from "firebase/auth";
import {
    collection,
    getDocs,
    onSnapshot,
    query,
    where,
} from "firebase/firestore";
import React, { useEffect, useState } from "react";
import { Bar } from "react-chartjs-2";

import { auth, db } from "@/util/firebase";

function BarChart() {
    const [authUser, setAuthUser] = useState(null);
    const [chartData, setChartData] = useState(null); // Changed variable name

    Chart.register(
        CategoryScale,
        LinearScale,
        PointElement,
        LineElement,
        Title,
        Tooltip,
        Legend,
        TimeScale,
        BarElement
    );

    useEffect(() => {
        onAuthStateChanged(auth, (user) => {
            user ? setAuthUser(user) : setAuthUser(null);
        });

        const fetchInformation = async () => {
            if (authUser) {
                const userId = authUser.uid;
                const q = query(
                    collection(db, "donors"),
                    where("restaurantId", "==", userId)
                );

                const querySnapshot = await getDocs(q);
                const chart = {};

                querySnapshot.forEach((doc) => {
                    const meal = doc.data();
                    const date = meal.date;

                    if (chart[date]) {
                        chart[date]++;
                    } else {
                        chart[date] = 1;
                    }
                });

                const chartArray = Object.keys(chart).map((date) => ({
                    date: date,
                    count: chart[date],
                }));
                const chartData = {
                    labels: chartArray.map((data) => data.date),
                    datasets: [
                        {
                            label: "Donation per day",
                            data: chartArray.map((data) => data.count),
                            backgroundColor: "rgba(75,192,192,1)",
                            borderColor: "rgba(0,0,0,1)",
                            borderWidth: 2,
                        },
                    ],
                    scales: {
                        x: {
                            type: "time",
                            time: {
                                unit: "day",
                            },
                        },
                    },
                };
                setChartData(chartData); // Corrected variable name
            }
        };

        const unsubscribe = onSnapshot(collection(db, "donors"), () => {
            fetchInformation();
        });

        return () => {
            unsubscribe();
        };
    }, [authUser]);

    return (
        <div className='w-full h-full'>
            {chartData && <Bar data={chartData} />}
        </div>
    );
}

export default BarChart;
