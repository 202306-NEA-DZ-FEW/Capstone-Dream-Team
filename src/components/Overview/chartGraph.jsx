import { onAuthStateChanged } from "firebase/auth";
import {
    collection,
    getDocs,
    onSnapshot,
    query,
    where,
} from "firebase/firestore";
import dynamic from "next/dynamic";
import { useTranslation } from "next-i18next";
import React, { useEffect, useState } from "react";

import { auth, db } from "@/util/firebase";

const ReactApexChart = dynamic(() => import("react-apexcharts"), {
    ssr: false,
});

function ChartGraph() {
    const [authUser, setAuthUser] = useState(null);
    const [chartData, setChartData] = useState(null);
    const { t } = useTranslation("common");
    useEffect(() => {
        onAuthStateChanged(auth, (user) => {
            user ? setAuthUser(user) : setAuthUser(null);
        });

        const fetchInformation = async () => {
            if (authUser) {
                const userId = authUser.uid;
                const q = query(
                    collection(db, "donors"),
                    where("restaurantId", "==", userId)
                );

                const querySnapshot = await getDocs(q);
                let chart = {};

                querySnapshot.forEach((doc) => {
                    const meal = doc.data();
                    // const dateParts = meal.date.split("/");
                    // const formattedDate = new Date(
                    //     dateParts[2],
                    //     dateParts[1] - 1,
                    //     dateParts[0]
                    // );
                    // const day = new Intl.DateTimeFormat("en-Us", {
                    //     weekday: "long",
                    // }).format(formattedDate);
                    const date = meal.date;
                    if (chart[date]) {
                        chart[date]++;
                    } else {
                        chart[date] = 1;
                    }
                });
                const chartArray = Object.keys(chart)
                    .map((date) => {
                        const dateParts = date.split("/");
                        const formattedDate = new Date(
                            dateParts[2],
                            dateParts[1] - 1,
                            dateParts[0]
                        );

                        return {
                            day: t(
                                "overview.chartGraph.weekdays." +
                                    formattedDate.getDay()
                            ),
                            // day: new Intl.DateTimeFormat("en-US", {
                            //     weekday: "long",
                            // }).format(formattedDate),
                            date: date,
                            count: chart[date],
                            formattedDate: formattedDate,
                        };
                    })
                    .sort((a, b) => a.formattedDate - b.formattedDate);

                // Add a null check before mapping
                if (chartArray && chartArray.length > 0) {
                    const chartData = {
                        series: [
                            {
                                name: "Donation per day",
                                data: chartArray.map((data) => data.count),
                            },
                        ],
                        options: {
                            stroke: {
                                curve: "smooth",
                            },

                            xaxis: {
                                type: "category",
                                categories: chartArray.map((data) => data.day),
                            },
                            yaxis: {
                                type: "numeric",
                                min: 0,
                                labels: {
                                    style: {
                                        colors: "#575757",
                                        fontSize: "14px",
                                    },
                                },
                                forceNiceScale: true,
                            },
                            tooltip: {
                                custom: function ({
                                    series,
                                    seriesIndex,
                                    dataPointIndex,
                                    w,
                                }) {
                                    const data = chartArray[dataPointIndex];
                                    return (
                                        `<div style="background-color: #ffffff; padding: 10px; border: 1px solid #cccccc; border-radius: 5px;">` +
                                        `<strong>${t(
                                            "overview.chartGraph.date"
                                        )}:</strong> ${data.formattedDate.toLocaleDateString()}<br>` +
                                        `<strong>${t(
                                            "overview.chartGraph.day"
                                        )}:</strong> ${data.day}<br>` +
                                        `<strong>${t(
                                            "overview.chartGraph.count"
                                        )}:</strong> ${data.count}</div>`
                                    );
                                },
                            },
                        },
                    };
                    setChartData(chartData);
                } else {
                    // Handle the case when chartArray is null or empty
                    setChartData(null);
                }
            }
        };

        const unsubscribe = onSnapshot(collection(db, "donors"), () => {
            fetchInformation();
        });

        return () => {
            unsubscribe();
        };
    }, [authUser]);

    return (
        <div className='col-span-12 rounded-md border border-stroke bg-white px-5 pt-7 pb-7 shadow sm:px-7.5 xl:col-span-8'>
            <div className='flex w-full flex-col flex-wrap items-start justify-between gap-3 sm:flex-nowrap'>
                <p className='font-bold text-primary'>
                    {t("overview.chartGraph.donationPerDay")}
                </p>
                <div className='w-full'>
                    <div className='-ml-5'>
                        {chartData && (
                            <ReactApexChart
                                options={chartData.options}
                                series={chartData.series}
                                type='area'
                                height={350}
                            />
                        )}
                    </div>
                </div>
            </div>
        </div>
    );
}

export default ChartGraph;
