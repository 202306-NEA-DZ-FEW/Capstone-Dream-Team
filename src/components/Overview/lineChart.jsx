import { Chart as ChartJS } from "chart.js/auto";
import React from "react";
import { Bar } from "react-chartjs-2";

function LineChart({ chartData }) {
    return <Bar data={chartData} />;
}

export default LineChart;
