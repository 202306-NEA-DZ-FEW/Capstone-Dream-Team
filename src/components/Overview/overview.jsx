import { useTranslation } from "next-i18next";
import React from "react";

import ChartGraph from "./chartGraph";
import StatCardList from "./statCardList";

export default function Overview() {
    const { t } = useTranslation("common");

    return (
        <div>
            <div className='mx-4 md:mx-6'>
                <div className='mt-4 pb-4 flex flex-col sm:flex-row items-center justify-between border-b border-gray-200 dark:border-gray-700'>
                    <div className='mb-2 sm:mb-0 tracking-wider font-light font-roboto'>
                        {t("history.History")}
                    </div>
                </div>
            </div>
            <StatCardList />
            <ChartGraph />
        </div>
    );
}
