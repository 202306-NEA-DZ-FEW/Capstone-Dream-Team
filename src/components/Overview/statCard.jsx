import React from "react";

export default function StatCard(props) {
    const { icon, title, number } = props;
    return (
        <div className='flex flex-col px-5 py-6 shadow-sm rounded-md border border-orange-100 bg-orange-50'>
            <div className='flex'>
                <div className='flex items-center h-11 w-11 justify-center p-3 rounded-full bg-orange-400 bg-opacity-80'>
                    <div className='text-2xl text-white'>{icon}</div>
                </div>

                <div className='mx-2'>
                    <h4 className='text-2xl font-bold text-gray-700 p-1'>
                        {number}
                    </h4>
                </div>
            </div>
            <div className='text-gray-500 overflow-hidden whitespace-nowrap text-base break-words pt-2'>
                {title}
            </div>
        </div>
    );
}
