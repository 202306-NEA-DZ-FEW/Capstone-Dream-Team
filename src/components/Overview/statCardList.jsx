import { onAuthStateChanged } from "firebase/auth";
import {
    collection,
    getDocs,
    onSnapshot,
    query,
    where,
} from "firebase/firestore";
import { useTranslation } from "next-i18next";
import React, { useEffect, useState } from "react";
import { BsPersonFillUp } from "react-icons/bs";
import { BsCurrencyDollar } from "react-icons/bs";
import { FaCheck } from "react-icons/fa6";
import { GiForkKnifeSpoon } from "react-icons/gi";

import { auth, db } from "@/util/firebase";

import StatCard from "./statCard";

const StatCardList = () => {
    const [authUser, setAuthUser] = useState(null);
    const [totalDonors, setTotalDonors] = useState(0);
    const [totalMeals, setTotalMeals] = useState(0);
    const [totalActiveMeals, setTotalActiveMeals] = useState(0);
    const [totalDonation, setTotalDonation] = useState(0);
    const { t } = useTranslation("common");
    useEffect(() => {
        // Define a function to fetch user information from Firestore.
        onAuthStateChanged(auth, (user) => {
            // If a user is authenticated, set 'authUser' to the user; otherwise, set it to null.
            user ? setAuthUser(user) : setAuthUser(null);
        });
        const fetchInformation = async () => {
            if (authUser) {
                const userId = authUser.uid; // Get the user's unique identifier (UID).
                const q = query(
                    collection(db, "donors"),
                    where("restaurantId", "==", userId)
                );
                const querySnapshot = await getDocs(q);
                setTotalDonors(querySnapshot.size);
                let totalMeals = 0;
                let totalActiveMeals = 0;
                let totalDonation = 0;
                querySnapshot.forEach((doc) => {
                    const meal = doc.data();
                    totalMeals += meal.quantity;
                    totalActiveMeals += meal.active_meal;
                    totalDonation += meal.quantity * meal.price;
                });
                setTotalMeals(totalMeals);
                setTotalActiveMeals(totalActiveMeals);
                setTotalDonation(totalDonation.toFixed(2));
            }
        };
        const unsubscribe = onSnapshot(collection(db, "donors"), () => {
            fetchInformation();
        });
        return () => {
            unsubscribe();
        };
    }, [authUser]);

    return (
        <div className='mx-6 mb-6'>
            <div className='flex flex-wrap -mx-6'>
                <div className='w-full mt-6 px-2 sm:w-1/2 xl:w-1/4 '>
                    <StatCard
                        icon={<GiForkKnifeSpoon />}
                        title={t("overview.statCard.totalMeals")}
                        number={totalMeals}
                    />
                </div>

                <div className='w-full mt-6 px-2 sm:w-1/2 xl:w-1/4 '>
                    <StatCard
                        icon={<BsPersonFillUp />}
                        title={t("overview.statCard.totalActiveMeals")}
                        number={totalActiveMeals}
                    />
                </div>
                <div className='w-full mt-6 px-2 sm:w-1/2 xl:w-1/4'>
                    <StatCard
                        icon={<FaCheck />}
                        title={t("overview.statCard.totalDonors")}
                        number={totalDonors}
                    />
                </div>
                <div className='w-full mt-6 px-2 sm:w-1/2 xl:w-1/4'>
                    <StatCard
                        icon={<BsCurrencyDollar />}
                        title={t("overview.statCard.totalDonation")}
                        number={totalDonation}
                    />
                </div>
            </div>
        </div>
    );
};

export default StatCardList;
