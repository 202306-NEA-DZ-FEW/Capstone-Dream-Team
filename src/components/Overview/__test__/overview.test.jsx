import renderer from "react-test-renderer";

import Overview from "../overview";
// Mock the child components
jest.mock("../chartGraph", () => ({ __esModule: true, default: jest.fn() }));
jest.mock("../statCardList", () => ({ __esModule: true, default: jest.fn() }));
it("renders correctly", () => {
    const tree = renderer.create(<Overview />).toJSON();
    expect(tree).toMatchSnapshot();
});
