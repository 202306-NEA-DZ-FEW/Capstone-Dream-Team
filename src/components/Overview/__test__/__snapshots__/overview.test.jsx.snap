// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div>
  <div
    className="mx-4 md:mx-6"
  >
    <div
      className="mt-4 pb-4 flex flex-col sm:flex-row items-center justify-between border-b border-gray-200 dark:border-gray-700"
    >
      <div
        className="mb-2 sm:mb-0 tracking-wider font-light font-roboto"
      >
        history.History
      </div>
    </div>
  </div>
</div>
`;
